### ISSUE TITLE
[Proposal] ElementStyles Registration

### ISSUE THREAD

**Issue by EisenbergEffect** [microsoft/fast-dna#2849]
## Description

This proposal is to add a `register` method to `ElementStyles`, which would allow the styles object to be registered into a global lookup by namespace and name. These styles could then later be retrieved by name and applied dynamically to the `adoptedStyleSheets` of a component.

The primary use case for this is in cases such as button, where we have the same component, but with several different styling options. If the customer only uses one of those options, we don't want them to have to pay for all of the options. We could handle this by creating different components, but since the only difference is styles, that would seem overly complicated. Simply being able to switch out styles seems a nicer solution. Another example of this is high-contrast only applying high-contrast styles if needed.

### Registering Styles

```TypeScript
import { css } from '@microsoft/fast-element';

export const styles = css`
  .foo { color: red; }
`.register('fast-button', 'lightweight')
```

The `css` tagged template helper returns an instance of `ElementStyles` to which we add a simple `register` method, which returns the element styles (enabling chaining). The `register` method will add the style object to a global registry, which can then be accessed elsewhere. 

### Finding Registered Styles

```TypeScript
import { StyleRegistry } from '@microsoft/fast-element';

const styles = StyleRegistry.find('fast-button', 'lightweight');
```

Any other component can lookup styles based on a known namespace and name.

### Adding/Removing Element Styles

```TypeScript
import { StyleRegistry, customElement, FastElement } from '@microsoft/fast-element';

@customElement({ ... })
export class FastButton extends FastElement {
  private appearanceStyles: ElementStyles | null = null;

  @attr appearance: string | null = null;
  appearanceChanged() {
    if (this.appearanceStyles !== null) {
      this.removeStyles(this.appearanceStyles);
    }

    this.appearanceStyles = StyleRegistry.find(this.tagName.toLowerCase(), 'lightweight');

    if (this.appearanceStyles !== null) {
      this.addStyles(this.appearanceStyles);
    }
  }
}
```

In this case, we'll add two methods to `FastElement`, implemented via `Controller` to enable adding and removing `ElementStyles` instances. As part of this, we'll also make some related changes to the `ElementStyles` interface to accommodate removal scenarios.

One final note is that with this implementation, end users would need to import any used optional styles in their JS code, to ensure that they are not tree-shaken out. There needs to be some static analyzable connection for bundlers. So, this allows us to decouple styles internally, in order to not force customers into using everything, but they do need to opt-in to the styles by importing them in their own code.

## Positive Side-Effect

* This would open up a new extensibility point for customers. If an element has something like an "appearance", then they can also write and register their own styles to be applied.
* With these APIs in place, the upcoming "recipes" feature could also use this to dynamically add different CSS as well.

## Futures

After the basics are in place, we could investigate lazy style loading. Imagine another API:

```TypeScript
const styles = await StyleRegistry.fetch(this.tagName.toLowerCase(), 'high-constrast');
```

In this case, a `fetch` request could be made to retrieve the styles and then rehydrate them into an `ElementStyles` instance, which could be added/removed using the above APIs.

**Comment by bheston** [comment]
This looks great. A registry type system is what I have been picturing as the most flexible representation of this reusable style idea.

Regarding the namespace, would it be required or validated in any way? It seems that many appearances will be common across components, especially the primitives notion of something like "accent plated" which would define something like a fill, border radius, and foreground color. This could be used for a button, selected list item, toggle switch, etc. I don't know if we need to provide any 'constant' types of namespaces or if someone could achieve something like this by using "shared" / "global" / etc. for the namespace.

**Comment by EisenbergEffect** [comment]
We could allow the namespace to be optional, and if not provided, put it in a "global" namespace. Rather than default that inside the API though, I might recommend that we manually use a known global namespace, such as "fast". Because I think customers may want to use this, I'd like to be careful that we don't end up with name collisions.

**Comment by chrisdholt** [comment]
@EisenbergEffect Does this change our export API in any additional way? 
```ts
export const CheckboxStyles = css`
    :host {}
`; <-- not registering because we don't need to

---

@customElement({
    name: "fast-checkbox",
    template,
    styles,
})
```
I'd assume this is additive and optional from an API standpoint; for instance, if a control doesn't need to support this it doesn't need to register in this way.

Regarding the name-spacing, I would agree to keep that out of the fast element implementation.  In `fast-components` I would expect we namespace any use case with `fast` as well to ensure no collisions. 

Regarding the implementation of this in our own system - I don't know that I would expect all controls to implement this behavior. I think one use case for all controls to implement this could be high contrast, but I want to be careful of assuming each component would benefit from first class support of this for varying appearances. I believe we still have robust support for folks wholesale swapping out styles and registering their own using the custom elements decorator and the pattern above. I tend to lean towards supporting opt-in behavior as a developer rather than opt-out. My initial thought on our adoption of this in `fast-components` would be that we offer this type of support as opt-in as well.

**Comment by EisenbergEffect** [comment]
This would be non-breaking, additive behavior in terms of how we're exporting and using our styles today. There's no requirement to register the styles. We would not just have the ability to do that if we have a need.

**Comment by EisenbergEffect** [comment]
Basically, we're just introducing a loose coupling mechanism for styles. Today, you could technical import various styles and manually apply them based on appearance values as well. It's a bit ugly and there's no way to remove them... So, this will improve the add/remove APIs as well as provide a loosely coupled lookup mechanism as well.

**Comment by chrisdholt** [comment]
> Basically, we're just introducing a loose coupling mechanism for styles. Today, you could technical import various styles and manually apply them based on appearance values as well. It's a bit ugly and there's no way to remove them... So, this will improve the add/remove APIs as well as provide a loosely coupled lookup mechanism.

@EisenbergEffect sure, I'm in agreement on the scoping and support here and the API's for `fast-element`. The second half of my statement was regarding how this applies in our own implementation of `fast-element` with our `fast-components`.

**Comment by EisenbergEffect** [comment]
Gotcha! Yeah, I don't think every control would use it, and once we add this capability, we'll probably need to do a few cycles to figure out how best to use it as well. I think there are some neat ways this could be combined with recipes and we'll probably have to tease those apart to figure out what goes where and how best to organize and apply things.

**Comment by nicholasrice** [comment]
This looks great!

**Comment by EisenbergEffect** [comment]
Adding to this proposal the idea of recipes. Something like this:

```TypeScript
import { css } from '@microsoft/fast-element';

export const styles = css`
  .foo { color: red; }
`.withRecipes([
    recipeA,
    recipeB,
    recipeC
  ]).register('fast-button', 'lightweight')
```

With this in place, then the `addStyles` API described above would add the style sheets as well as connect the recipes.

I think the recipes would just have a bind/unbind so that they could participate in the behavior lifecycle of the element's view. In this way, they could even contain behavior, such as the code for a material ripple effect (should we call this `withBehavior` instead?).

Question for @nicholasrice Do the recipes themselves need to maintain state or do they just need to be applied against the host element (or its shadow dom)?

**Comment by nicholasrice** [comment]
@EisenbergEffect not currently, and I don't anticipate them needing to maintain state. They're just a function of the design-system data.

**Comment by EisenbergEffect** [comment]
Implemented in #2925
